File: js/movement/geometry.js
/* ─── Movement Assessment — pure 2D geometry ───────────────────────────────
 *
 * Image-space geometry only. Nothing here knows about squats, sides or
 * providers, and nothing here is a 3D or anatomical joint angle: these are
 * angles between PROJECTED landmark positions in one camera image.
 * ───────────────────────────────────────────────────────────────────────── */

/* Segments shorter than this (in the same units as the points — image pixels
   in practice) have no direction, so an angle at them is undefined. */
export const MIN_SEGMENT_LENGTH = 1e-9;

function isPoint(p) {
  return !!p && typeof p.x === "number" && typeof p.y === "number" &&
    Number.isFinite(p.x) && Number.isFinite(p.y);
}

/* Included angle at B formed by A–B–C, in DEGREES, range [0, 180].
 *
 *   u = A − B,  v = C − B
 *   angle = atan2(|u × v|, u · v)
 *
 * atan2 of the cross and dot products is used instead of acos(u·v / |u||v|)
 * because it is exact at 0° and 180° (acos loses precision there and can
 * return NaN when rounding pushes its argument past ±1).
 *
 * 180° means A, B and C are collinear with B between them (a straight limb);
 * smaller values mean more flexion at B. The result is unsigned: the 2D
 * projection cannot say which way the joint bends.
 *
 * Returns null — never NaN or Infinity — when any point is missing or
 * non-finite, or when A or C coincides with B. */
export function includedAngleDeg(a, b, c) {
  if (!isPoint(a) || !isPoint(b) || !isPoint(c)) return null;
  const ux = a.x - b.x, uy = a.y - b.y;
  const vx = c.x - b.x, vy = c.y - b.y;
  const lenU = Math.hypot(ux, uy);
  const lenV = Math.hypot(vx, vy);
  if (!Number.isFinite(lenU) || !Number.isFinite(lenV)) return null;
  if (lenU < MIN_SEGMENT_LENGTH || lenV < MIN_SEGMENT_LENGTH) return null;
  const cross = ux * vy - uy * vx;
  const dot = ux * vx + uy * vy;
  const deg = Math.atan2(Math.abs(cross), dot) * 180 / Math.PI;
  return Number.isFinite(deg) ? deg : null;
}

export function distance(a, b) {
  if (!isPoint(a) || !isPoint(b)) return null;
  const d = Math.hypot(a.x - b.x, a.y - b.y);
  return Number.isFinite(d) ? d : null;
}

export function midpoint(a, b) {
  if (!isPoint(a) || !isPoint(b)) return null;
  return { x: (a.x + b.x) / 2, y: (a.y + b.y) / 2 };
}

/* Median of finite numbers (mean of the two middle values for an even
   count). Non-finite entries are ignored; an empty input gives null. */
export function median(values) {
  const v = (values || []).filter(function (x) { return typeof x === "number" && Number.isFinite(x); })
    .sort(function (p, q) { return p - q; });
  if (v.length === 0) return null;
  const mid = Math.floor(v.length / 2);
  return v.length % 2 === 1 ? v[mid] : (v[mid - 1] + v[mid]) / 2;
}

/* Mean of finite numbers, or null. Values are summed in sorted order so the
   result does not depend on input order (floating-point addition is not
   associative). */
export function mean(values) {
  const v = (values || []).filter(function (x) { return typeof x === "number" && Number.isFinite(x); })
    .sort(function (p, q) { return p - q; });
  if (v.length === 0) return null;
  return v.reduce(function (s, x) { return s + x; }, 0) / v.length;
}

/* Nearest-rank percentile (p in [0, 100]) of finite numbers, or null. */
export function percentile(values, p) {
  const v = (values || []).filter(function (x) { return typeof x === "number" && Number.isFinite(x); })
    .sort(function (a, b) { return a - b; });
  if (v.length === 0) return null;
  const rank = Math.ceil((p / 100) * v.length);
  return v[Math.min(v.length - 1, Math.max(0, rank - 1))];
}

File: research_backend/physiq_research/domain/geometry.py
"""Pure 2D image-space geometry (port of js/movement/geometry.js).

Angles here are between PROJECTED landmark positions in one camera image;
nothing is a 3D or anatomical joint angle.
"""

from __future__ import annotations

import math
from collections.abc import Iterable
from typing import Protocol

# Segments shorter than this (image pixels in practice) have no direction.
MIN_SEGMENT_LENGTH = 1e-9


class PointLike(Protocol):
    @property
    def x(self) -> float: ...

    @property
    def y(self) -> float: ...


class Point:
    __slots__ = ("x", "y")

    def __init__(self, x: float, y: float) -> None:
        self.x = x
        self.y = y

    def __repr__(self) -> str:  # pragma: no cover - debugging aid only
        return f"Point({self.x!r}, {self.y!r})"


def is_finite_number(value: object) -> bool:
    """True for a finite real number (bool is not a number here)."""
    if isinstance(value, bool):
        return False
    if isinstance(value, (int, float)):
        return math.isfinite(value)
    return False


def _is_point(p: object) -> bool:
    if p is None:
        return False
    x = getattr(p, "x", None)
    y = getattr(p, "y", None)
    return is_finite_number(x) and is_finite_number(y)


def included_angle_deg(a: PointLike | None, b: PointLike | None, c: PointLike | None) -> float | None:
    """Included angle at B formed by A–B–C, degrees, range [0, 180].

    u = A − B, v = C − B, angle = atan2(|u × v|, u · v). Exact at 0° and
    180° (unlike acos). Returns None for missing/non-finite points or when A
    or C coincides with B.
    """
    if not (_is_point(a) and _is_point(b) and _is_point(c)):
        return None
    assert a is not None and b is not None and c is not None
    ux, uy = a.x - b.x, a.y - b.y
    vx, vy = c.x - b.x, c.y - b.y
    len_u = math.hypot(ux, uy)
    len_v = math.hypot(vx, vy)
    if not (math.isfinite(len_u) and math.isfinite(len_v)):
        return None
    if len_u < MIN_SEGMENT_LENGTH or len_v < MIN_SEGMENT_LENGTH:
        return None
    cross = ux * vy - uy * vx
    dot = ux * vx + uy * vy
    deg = math.atan2(abs(cross), dot) * 180 / math.pi
    return deg if math.isfinite(deg) else None


def distance(a: PointLike | None, b: PointLike | None) -> float | None:
    if not (_is_point(a) and _is_point(b)):
        return None
    assert a is not None and b is not None
    d = math.hypot(a.x - b.x, a.y - b.y)
    return d if math.isfinite(d) else None


def midpoint(a: PointLike | None, b: PointLike | None) -> Point | None:
    if not (_is_point(a) and _is_point(b)):
        return None
    assert a is not None and b is not None
    return Point((a.x + b.x) / 2, (a.y + b.y) / 2)


def _finite_sorted(values: Iterable[object]) -> list[float]:
    return sorted(float(v) for v in values if is_finite_number(v))  # type: ignore[arg-type]


def median(values: Iterable[object]) -> float | None:
    """Median of the finite numbers (mean of the middle two for an even count)."""
    v = _finite_sorted(values)
    if not v:
        return None
    mid = len(v) // 2
    return v[mid] if len(v) % 2 == 1 else (v[mid - 1] + v[mid]) / 2


def mean(values: Iterable[object]) -> float | None:
    """Mean of the finite numbers, summed in sorted order (order-independent)."""
    v = _finite_sorted(values)
    if not v:
        return None
    total = 0.0
    for x in v:
        total += x
    return total / len(v)


def percentile(values: Iterable[object], p: float) -> float | None:
    """Nearest-rank percentile (p in [0, 100]) of the finite numbers."""
    v = _finite_sorted(values)
    if not v:
        return None
    rank = math.ceil((p / 100) * len(v))
    return v[min(len(v) - 1, max(0, rank - 1))]
